ITISFoundation/osparc-simcore
# bugfix for #1853: garbage collector deletes guest users

## What do these changes do?

Access to ``/study/{study_id}`` link produces
![image](https://user-images.githubusercontent.com/32402063/98027429-f132a100-1e0c-11eb-84ac-67914f1335b5.png)

The reason is that the Garbage collector removes GUEST users when they have no resources associated. This caused that temporary GUEST users created in studies access were deleted even before their projects were associated.

This fix overcomes this problem by 
- first creating the user as ANONYMOUS instead of GUEST
- then, once the ``access_study`` is processed, a fictitious socket is created for the socket and it is turn into GUEST

This will prevent the GC to destroy the user before the ``access_study`` handle is completed and some extra time TTL to reconnect.

- [x] Adapted ``pytest tests/unit/with_dbs/fast/test_access_to_studies.py``
- [x] test to demo that an eager GC does not delete user
- cleaned

## Related issue number


Fixes #1853

## How to test
```cmd
cd services/web
make install-dev
pytest --pdb -s --log-cli-level=DEBUG tests/unit/with_dbs/fast/test_access_to_studies.py
```
and can also see GC log.

## Checklist
